### ISSUE TITLE
[HOLD for payment 2024-06-11] [$1000] Web - Three dot menu is not opened via Enter key in Request Money to Add Receipt

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#33502]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.16-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Login to ND Expensify
2. Tap FAB button
3. Tap Request money
4. Select Manual tab
5. Select a currency and enter any amount
6. Tap Next
7. On the Payor screen
8. Tab until you get to the "three dot menu"
9. Press `Enter` key

## Expected Result:
The "Add receipt" option should pop up after pressing `Enter` key on three dot menu

## Actual Result:
Nothing happens when press Enter key while the focus on the three dot menu in Manual Expense modal

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/a35505e2-80f6-4d78-96f0-1b7b4407172a

![Bug6323248_1703196651854!ellipsis](https://github.com/Expensify/App/assets/78819774/9fc81838-d15c-4978-ad32-f73b292a4e69)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d30724d2128b5dca</li>
        <li>Upwork Job ID: 1738244084031193088</li>
        <li>Last Price Increase: 2024-01-23</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Christinadobrzyn</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Three dot menu is not opened via Enter key in Request Money to Add Receipt

### What is the root cause of that problem?

### What changes do you think we should make in order to solve the problem?

PressableWithFeedback lacks the accessible `true` prop
### What alternative solutions did you explore? (Optional)
We should pass accesible prop to `PressableWithFeedback`

**Comment by situchan** [comment]
This is accessibility issue and seems like not a fit to any of waves

**Comment by Christinadobrzyn** [comment]
I think this might fit with Wave 6 - although it might be a low priority.

**Comment by Christinadobrzyn** [comment]
Talking this over in chat with the team - we should bring this flow up to parity with `Tab` & `Enter` on the rest of the app. So that means that you should be able to `Tab` to the three dots and click `Enter` to open the "add receipt" window (a similar behaviour to other parts of the app). However, you don't need to start the Add Receipt process by clicking Enter. 

So I think this should be worked on. I think it can be external. 

@mollfpr feel free to review any proposals.

**Comment by FitseTLT** [comment]
Updated [Proposal](https://github.com/Expensify/App/issues/33502#issuecomment-1867906905
)

**Comment by mollfpr** [comment]
Taking a review on the proposal now!

**Comment by Christinadobrzyn** [comment]
Hi @mollfpr how are the proposals looking?

**Comment by Christinadobrzyn** [comment]
I'm going ooo for a week so going to assign a buddy to monitor this for me. 

Status - we're reviewing proposals.

**Comment by laurenreidexpensify** [comment]
@mollfpr pls confirm if you have reviewed proposals thank you

**Comment by laurenreidexpensify** [comment]
Bump @mollfpr thanks

**Comment by FitseTLT** [comment]
@mollfpr on Pressing enter it is calling the `onPress` for the second time for `keyup` ( for releasing the enter key too) event so we avoid running the callback for the second time for it 👍

**Comment by mollfpr** [comment]
@FitseTLT Why it's call for the second time? Is it because the listener in the `BaseOptionsSelector`?

**Comment by mollfpr** [comment]
I don't think we have the same issue with the linked issue upstream. We don't trigger the disable state on press.

**Comment by bernhardoj** [comment]
> the issue above is specific to disabled button, but the real cause is the button is blurred

☝️

**Comment by FitseTLT** [comment]
We may also use `preventDefault` may be to prevent the blurring inside `onPress` but I think checking the event type `keyup` approach is better.

**Comment by mollfpr** [comment]
Do we have the same issue with another component?

**Comment by FitseTLT** [comment]
Not that I know of.

**Comment by bernhardoj** [comment]
It happens with every pressable with a role of button

**Comment by laurenreidexpensify** [comment]
@bernhardoj do you think we're any closer to a solution with the progress above?

**Comment by bernhardoj** [comment]
I think not, this should be fixed on the upstream package (except we want to accept a workaround).

**Comment by mollfpr** [comment]
Ideally, the upstream fix will be preferred; we want to fix it fast, and then the workaround is fine, but we should have a better workaround to fix the same issue that may happen in another component.

**Comment by FitseTLT** [comment]
We can apply the workaround in the Pressable component when the role is button

**Comment by laurenreidexpensify** [comment]
Still waiting for a solution, not overdue, active conversation continues

**Comment by mollfpr** [comment]
I think we should have the fix for upstream because this issue is not a priority for now.

**Comment by laurenreidexpensify** [comment]
@mollfpr can you write a few lines of summary of what you suggest happens next on this one? thanks!

**Comment by mollfpr** [comment]
@laurenreidexpensify I'm still inclined to have the issue fixed on RNW upstream. Either way, we keep the issue open and double the bounty or accept the workaround (depending on the urgency).

**Comment by Christinadobrzyn** [comment]
I'll increase the bounty so we can try and not use a workaround as a solution. We are collecting proposals for this

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The popover menu is open and then closed immediately when we open it with the Enter key.

### What is the root cause of that problem?
This is a bug in react native web. The root cause can be found on the upstream [PR](https://www.github.com/necolas/react-native-web/pull/2632) I open.

### What changes do you think we should make in order to solve the problem?

Fix it upstream. I have opened a [PR](https://www.github.com/necolas/react-native-web/pull/2632) here waiting for a review.

**Comment by mollfpr** [comment]
I tested the fix upstream by @bernhardoj, and it looks good. We still need feedback from the maintainer, but I think it's on the right path.

Let's go with @bernhardoj [proposal](https://github.com/Expensify/App/issues/33502#issuecomment-1907604187)!

🎀 👀 🎀 C+ reviewed!

**Comment by neil-marcellini** [comment]
> I tested the fix upstream by @bernhardoj, and it looks good. We still need feedback from the maintainer, but I think it's on the right path.
> 
> Let's go with @bernhardoj [proposal](https://github.com/Expensify/App/issues/33502#issuecomment-1907604187)!
> 
> 🎀 👀 🎀 C+ reviewed!

Ok sounds good. I'll go ahead and assign him. However, I'm not sure if we care about fixing keyboard bugs right now, so I have asked in Slack (internally) [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1706125987236429). We might close this. It will be great to fix react native web of course, either way. cc @Christinadobrzyn

**Comment by neil-marcellini** [comment]
Closing since we plan to solve keyboard navigation issues all at once as part of a project, hopefully soon.

**Comment by bernhardoj** [comment]
@neil-marcellini hi, the previous decision [here](https://github.com/Expensify/App/issues/33502#issuecomment-1868138280) was to fix this particular issue which leads to all discussion and finding the non-workaround solution. I would really appreciate if we can stick with the decision 😊. The upstream PR is ready (even though not reviewed yet by the maintainer) and eventually we need to fix this sooner or later (sooner is better as this is affecting all button with popover). 

Most of the work will be done on the upstream, so we can keep this hold weekly and we can still focus on other issue and only come back here once the upstream PR is merged to upgrade the rn-web version.

TIA!

cc: @Christinadobrzyn @mollfpr

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Enter key after tab navigation will not open the menu.

### What is the root cause of that problem?

Coming from here https://github.com/Expensify/App/issues/37464

After debugging the issue, I noticed that the Button component trigger action twice with every ENTER key from keyboard.
After fixing the issue here https://github.com/Expensify/App/issues/37464#issuecomment-1971486442
I am here posting a proposal to fix this issue.

Due to the update on the [propsal](https://github.com/Expensify/App/issues/37464#issuecomment-1971486442) the button component is now working perfectly and does not trigger twice.

The issue whith the 3 dots is that the key ENTER from keyboard trigger twice, the first one show the PopoverMenu and the next second action hide it.

### What changes do you think we should make in order to solve the problem?

Change the component `PressableWithoutFeedback` with the button component. The button component is more enhanced to check keyboard event and to prevent twice action (after change on my proposal)

```diff
- <PressableWithoutFeedback
+ <Button
  onPress={() => {
    if (isPopupMenuVisible) {
    hidePopoverMenu();
```

POC:

https://github.com/Expensify/App/assets/12425932/b22b0ea1-a00c-4594-b856-ec04837b90ce


### What alternative solutions did you explore? (Optional)

**Comment by roryabraham** [comment]
Going to reopen this since @bernhardoj has prepared an upstream fix which has been merged.

**Comment by roryabraham** [comment]
@bernhardoj please proceed with a PR to upgrade react-native-web to include your upstream fix.

**Comment by roryabraham** [comment]
I understand why @neil-marcellini closed this, but given that @bernhardoj was hired for the [upstream PR](https://github.com/necolas/react-native-web/pull/2632), and that PR has been merged. So I think we should let him finish this – he already did a good portion of the work for which he was hired.

**Comment by bernhardoj** [comment]
Thanks, preparing the PR...

**Comment by bernhardoj** [comment]
PR is ready

cc: @mollfpr @roryabraham

**Comment by Christinadobrzyn** [comment]
hi @mollfpr @bernhardoj just checking in on this - can you provide an update? TY!

**Comment by bernhardoj** [comment]
Hi, the update is in this [comment](https://github.com/Expensify/App/issues/26990#issuecomment-2084286201). We are currently blocked by an upstream issue.

**Comment by Christinadobrzyn** [comment]
watching - https://github.com/Expensify/App/issues/26990#issuecomment-2084286201

**Comment by Christinadobrzyn** [comment]
monitoring - https://github.com/Expensify/App/issues/26990#issuecomment-2109507227

**Comment by Christinadobrzyn** [comment]
It looks like https://github.com/necolas/react-native-web/pull/2601 was closed @bernhardoj do you think we can start working on this again?

**Comment by bernhardoj** [comment]
@Christinadobrzyn @mollfpr already approved the PR, waiting for @neil-marcellini review

**Comment by Christinadobrzyn** [comment]
Payouts due:

- [ ] Contributor: $1000 @bernhardoj (Upwork offer - https://www.upwork.com/nx/wm/offer/102682328)
- [ ] Contributor+: $1000 @mollfpr (via NewDot payment)

Upwork job is [here](https://www.upwork.com/jobs/~014cd511c8399325b2).

Even though this is Internal - do we want a regression test?

**Comment by mollfpr** [comment]
>[@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR. The fix is upstream.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression step should be enough.

> [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

1. Open the money request page
2. Enter any amount
3. Select any participant
4. In the confirmation page, keep press TAB until the three-dot menu is highlighted
5. Press ENTER
6. **_Verify_** the menu is shown
7. 👍 or 👎

**Comment by JmillsExpensify** [comment]
$1,000 approved for @mollfpr

**Comment by Christinadobrzyn** [comment]
Regression test- https://github.com/Expensify/Expensify/issues/404150

Paid out based on [Payment Summary ](https://github.com/Expensify/App/issues/33502#issuecomment-2161332155)

Closing!
